src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs

```diff
@@ -183,6 +184,41 @@ public static string Utf8GetString(ReadOnlySpan<byte> bytes)
 #endif
         }
 
+        public static bool TryLookupUtf8Key<TValue>(
+            this Dictionary<string, TValue> dictionary,
+            ReadOnlySpan<byte> utf8Key,
+            [MaybeNullWhen(false)] out TValue result)
+        {
+#if NET9_0_OR_GREATER
+            Debug.Assert(dictionary.Comparer is IAlternateEqualityComparer<ReadOnlySpan<char>, string>);
+
+            Dictionary<string, TValue>.AlternateLookup<ReadOnlySpan<char>> spanLookup =
+                dictionary.GetAlternateLookup<string, TValue, ReadOnlySpan<char>>();
```

Comment:
Do you control the dictionary here? Would it make sense to use a custom comparer that supports direct comparison between UTF8 and UTF16 keys?

Reply:
I do, but depending on configuration this could be using case insensitive comparison which complicates concerns. Having such a comparer baked into `StringComparer` (e.g. obtained via a factory accepting an `Encoding`) would certainly help though.

Reply:
The other issue (that I think we were discussing when alternate lookup was being implemented) is that a `ReadOnlySpan<byte>` to `string` alternate lookup would necessitate decoding the LHS on every hashcode computation and subsequent equality comparison(s).